transports/quic/src/transport.rs

```diff
@@ -198,11 +206,67 @@ impl<P: Provider> Transport for GenTransport<P> {
         &mut self,
         addr: Multiaddr,
     ) -> Result<Self::Dial, TransportError<Self::Error>> {
-        // TODO: As the listener of a QUIC hole punch, we need to send a random UDP packet to the
-        // `addr`. See DCUtR specification below.
-        //
-        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
-        Err(TransportError::MultiaddrNotSupported(addr))
+        let (socket_addr, _version, peer_id) =
+            multiaddr_to_socketaddr(&addr, self.support_draft_29)
+                .ok_or_else(|| TransportError::MultiaddrNotSupported(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+
+        let Some(peer_id) = peer_id else {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        };
+
+        let listeners = self
+            .listeners
+            .iter()
+            .filter(|l| {
+                if l.is_closed {
+                    return false;
+                }
+                let listen_addr = l.endpoint_channel.socket_addr();
+                SocketFamily::is_same(&listen_addr.ip(), &socket_addr.ip())
+                    && listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()
+            })
+            .collect::<Vec<_>>();
+
+        // which listener to use to send packets? go-libp2p uses routing table with random port
+        let endpoint_channel = match listeners.len() {
+            0 => {
+                return Err(TransportError::Other(
+                    Error::NoActiveListenerForDialAsListener,
+                ));
+            }
+            1 => listeners[0].endpoint_channel.clone(),
+            _ => {
+                // Pick any listener to use for dialing.
+                // We hash the socket address to achieve determinism.
+                let mut hasher = DefaultHasher::new();
+                socket_addr.hash(&mut hasher);
+                let index = hasher.finish() as usize % listeners.len();
+                listeners[index].endpoint_channel.clone()
+            }
+        };
+
+        let hole_puncher = HolePuncher::new(endpoint_channel, socket_addr, self.handshake_timeout);
+
+        let (sender, receiver) = oneshot::channel();
+
+        self.hole_punch_map
+            .lock()
+            .unwrap()
+            .insert((socket_addr, peer_id), sender);
```

Comment:
What's the logic should be if we try to hole punch `(socket_addr, peer_id)` more than one time?

Reply:
[go-libp2p](https://github.com/libp2p/go-libp2p/blob/a4321de73f731f90925faf3df7150bfb4ab70e15/p2p/transport/quic/transport.go#L204-L207) returns an error. We can do the same with say a new error type `HolePunchInProgress(SocketAddr, PeerId)`?

Reply:
I think so